File: web/src/CredentialService.ts
import ServiceBase from "./ServiceBase";
import {
  CheckStatusRequest,
  CheckStatusResponse,
  CreateProofRequest,
  CreateProofResponse,
  IssueFromTemplateRequest,
  IssueFromTemplateResponse,
  IssueRequest,
  IssueResponse,
  SendRequest,
  SendResponse,
  ServiceOptions,
  UpdateStatusRequest,
  UpdateStatusResponse,
  VerifiableCredentialDefinition,
  VerifyProofRequest,
  VerifyProofResponse,
} from "./proto";

import type {Client as BrowserClient} from "nice-grpc-web";

export class CredentialService extends ServiceBase {
  client: BrowserClient<typeof VerifiableCredentialDefinition>;

  constructor(options?: ServiceOptions) {
    super(options);

    this.client = this.createClient(
      VerifiableCredentialDefinition
    );
  }

  public async issueCredential(request: IssueRequest): Promise<IssueResponse> {
    return this.client.issue(request, {
      metadata: await this.getMetadata(IssueRequest.encode(request).finish()),
    });
  }

  public async issueFromTemplate(
    request: IssueFromTemplateRequest
  ): Promise<IssueFromTemplateResponse> {
    return this.client.issueFromTemplate(request, {
      metadata: await this.getMetadata(
        IssueFromTemplateRequest.encode(request).finish()
      ),
    });
  }

  public async createProof(
    request: CreateProofRequest
  ): Promise<CreateProofResponse> {
    return this.client.createProof(request, {
      metadata: await this.getMetadata(
        CreateProofRequest.encode(request).finish()
      ),
    });
  }

  public async verifyProof(
    request: VerifyProofRequest
  ): Promise<VerifyProofResponse> {
    return this.client.verifyProof(request, {
      metadata: await this.getMetadata(
        VerifyProofRequest.encode(request).finish()
      ),
    });
  }

  public async checkStatus(
    request: CheckStatusRequest
  ): Promise<CheckStatusResponse> {
    return this.client.checkStatus(request, {
      metadata: await this.getMetadata(
        CheckStatusRequest.encode(request).finish()
      ),
    });
  }

  public async updateStatus(
    request: UpdateStatusRequest
  ): Promise<UpdateStatusResponse> {
    return this.client.updateStatus(request, {
      metadata: await this.getMetadata(
        UpdateStatusRequest.encode(request).finish()
      ),
    });
  }

  public async send(request: SendRequest): Promise<SendResponse> {
    return this.client.send(request, {
      metadata: await this.getMetadata(SendRequest.encode(request).finish()),
    });
  }
}

File: java/src/main/java/trinsic/services/CredentialService.java
package trinsic.services;

import com.google.common.util.concurrent.ListenableFuture;
import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.DidException;
import trinsic.sdk.options.v1.Options;
import trinsic.services.verifiablecredentials.v1.*;

import java.util.concurrent.ExecutionException;

public class CredentialService extends ServiceBase {
  private final VerifiableCredentialGrpc.VerifiableCredentialFutureStub stub;

  public CredentialService() {
    this(null);
  }

  public CredentialService(Options.ServiceOptions.Builder options) {
    super(options);

    this.stub = VerifiableCredentialGrpc.newFutureStub(this.getChannel());
  }

  public ListenableFuture<IssueResponse> issueCredential(
      IssueRequest request)
      throws InvalidProtocolBufferException, DidException {
    if (request.getDocumentJson().isBlank())
      throw new IllegalArgumentException("document json must not be empty");

    return withMetadata(stub, request).issue(request);
  }

  public ListenableFuture<IssueFromTemplateResponse>
      issueCredentialFromTemplate(IssueFromTemplateRequest request)
          throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).issueFromTemplate(request);
  }

  public ListenableFuture<CheckStatusResponse> checkStatus(
      CheckStatusRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).checkStatus(request);
  }

  public void updateStatus(UpdateStatusRequest request)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    var response = withMetadata(stub, request).updateStatus(request).get();
  }

  public ListenableFuture<CreateProofResponse> createProof(
      CreateProofRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).createProof(request);
  }

  public ListenableFuture<VerifyProofResponse> verifyProof(
      VerifyProofRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).verifyProof(request);
  }

  public void send(SendRequest request)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    var response = withMetadata(stub, request).send(request).get();
  }
}
